agent_api/src/main/java/dev/aikido/agent_api/storage/service_configuration/ParsedFirewallLists.java

```diff
@@ -0,0 +1,141 @@
+
+public class ParsedFirewallLists {
+    private final List<IPEntry> blockedIps = new ArrayList<>();
+    private final List<IPEntry> allowedIps = new ArrayList<>();
+    private final List<UADetailsEntry> uaDetails = new ArrayList<>();
+    private Pattern blockedUserAgents = null;
+    private Pattern monitoredUserAgents = null;
+
+    public ParsedFirewallLists() {
+
+    }
+
+    public List<Match> matchBlockedIps(String ip) {
+        List<Match> matches = new ArrayList<>();
+        for (IPEntry entry : this.blockedIps) {
+            if (entry.ips().matches(ip)) {
+                matches.add(new Match(entry.key(), !entry.monitor(), entry.description()));
+            }
+        }
+        return matches;
+    }
+
+    // returns true if one or more matches has been found with allowlist.
+    public boolean matchesAllowedIps(String ip) {
+        if (this.allowedIps.isEmpty()) {
+            return true; // Empty allowed is means all ips match
+        }
+        for (IPEntry entry : this.allowedIps) {
+            if (entry.ips().matches(ip)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public UABlockedResult matchBlockedUserAgents(String userAgent) {
+        boolean isBlocked = false;
+        if (blockedUserAgents != null)
+            isBlocked = blockedUserAgents.matcher(userAgent).find();
+
+        boolean isMonitored = false;
+        if (monitoredUserAgents != null)
+            isMonitored = monitoredUserAgents.matcher(userAgent).find();
+
+        if (!isMonitored && !isBlocked)
+            // only run the more detailed matches if it's an actual attack/monitored.
+            return new UABlockedResult(false, List.of());
+
+        List<String> matchedUAKeys = new ArrayList<>();
+        for (UADetailsEntry entry : this.uaDetails) {
+            if (entry.pattern().matcher(userAgent).find()) {
+                matchedUAKeys.add(entry.key());
+            }
+        }
+        return new UABlockedResult(isBlocked, matchedUAKeys);
+    }
+
+    public void update(ReportingApi.APIListsResponse response) {
+        updateBlockedIps(response.blockedIPAddresses());
+        updateMonitoredIps(response.monitoredIPAddresses());
+        updateAllowedIps(response.allowedIPAddresses());
+
+        updateBlockedAndMonitoredUAs(response.blockedUserAgents(), response.monitoredUserAgents());
+        updateUADetails(response.userAgentDetails());
+    }
+
+    public void updateBlockedIps(List<ReportingApi.ListsResponseEntry> blockedIpsList) {
+        blockedIps.clear();
+        if (blockedIpsList == null)
+            return;
+        for (ReportingApi.ListsResponseEntry entry : blockedIpsList) {
+            IPList ipList = createIPList(entry.ips());
+            blockedIps.add(new IPEntry(/* monitor */ false, entry.key(), entry.source(), entry.description(), ipList));
+        }
+    }
+
+    public void updateMonitoredIps(List<ReportingApi.ListsResponseEntry> monitoredIpsList) {
```

Comment:
I was looking for a `.clear()` here but it happens in `updateBlockedIps`, not very obvious

What if someone re-orders these lines? 🔥 

```java
updateBlockedIps(response.blockedIPAddresses());
updateMonitoredIps(response.monitoredIPAddresses());
```

Reply:
Perhaps better to split off in `monitoredIps` and to clear on its own?

Reply:
```suggestion
    public void updateMonitoredIps(List<ReportingApi.ListsResponseEntry> monitoredIpsLists) {
```

Reply:
yep I know, was just easiest to add the updateMonitoredIps that way, will change